xclib/examples/fast_evaluate.py

```diff
@@ -0,0 +1,79 @@
+import argparse
+import numpy as np
+from xclib.evaluation.xc_metrics import _get_topk, precision, recall, psprecision, compute_inv_propesity, calc_gt_metrics
+import scipy.sparse as sp
+
+class XCEvaluator():
+    def __init__(self, args):
+        super(XCEvaluator, self).__init__()
+        self.args = args
+        self.k = 200
+        self.top_indices = None
+        self.smat = None
+        self.tst_X_Y = sp.load_npz(args.tst_lbl_pth)
+        self.trn_X_Y = sp.load_npz(args.trn_lbl_pth)
+        self.A = args.A
+        self.B = args.B
+    
+    def process_indices(self):
+        '''
+        Ensure that there are no duplicate indices for each data point _get_topk will pad with the index num_labels for remaining slots, this will be problematic in in1d later
+        '''
+        num_docs, num_labels = self.tst_X_Y.shape
+        offsets = np.arange(self.k)
+        padded_indices = np.where(self.top_indices == num_labels)
+        offset_arr = np.zeros((num_docs, self.k))
+        
+        offset_arr[padded_indices] = 1
+        self.top_indices = self.top_indices + offset_arr * offsets
+        
+    def compute_top_indices(self):
+        self.smat = sp.load_npz(args.smat_pth)
+        if args.top_indices_pth is not None:
+            self.top_indices = np.load(args.top_indices_pth)
+        else:
+            self.top_indices = _get_topk(self.smat, k = self.k)
+        self.process_indices()
+    
+    def build_metrics_dict(self, precs, recalls, psps, recall_at_gt, micro_recall_at_gt):
+        ks = [1, 3, 5, 10, 50, 100]
+        metrics_dict = {}
+        for k in ks:
+            metrics_dict[f'P@{k}'] = f'{precs[k - 1] :.5f}'
+            metrics_dict[f'R@{k}'] = f'{recalls[k - 1] :.5f}'
+            metrics_dict[f'PSP@{k}'] = f'{psps[k - 1] :.5f}'
+        metrics_dict['R@GT'] = f'{recall_at_gt :.5f}'
+        metrics_dict['MicroR@GT'] = f'{micro_recall_at_gt :.5f}'
+        return metrics_dict
+    
+    def compute_metrics(self):
```

Comment:
We need a better way to supply k. People compute recall at higher values (10/20/50/100 etc.,) as compared to precision (1-5). BTW, nDCG is missing from the example.

What do you think?

Reply:
- k is now passed as a command line argument (refer [here](https://github.com/kunaldahiya/pyxclib/blob/fc3bb71c57283ea28ef8f9377572534102a10918/xclib/examples/fast_evaluate.py#L57)). 
- nDCG has now been added (refer [here](https://github.com/kunaldahiya/pyxclib/blob/DeepakSaini119/fast_metrics/xclib/examples/fast_evaluate.py#L39)
- Although metrics like P@100 will be very low for the datasets we use, we wanted to verify for a particular `k` is the behavior across metrics the same.